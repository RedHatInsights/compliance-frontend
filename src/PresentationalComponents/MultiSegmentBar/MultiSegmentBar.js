import {
  Chart,
  ChartBar,
  ChartContainer,
  ChartStack,
  ChartTooltip,
} from '@patternfly/react-charts';
import propTypes from 'prop-types';
import React, { useState, useEffect, useRef } from 'react';

const tooltipXOffset = (total, width, count) => -((width * count) / total / 2);

const MultiSegmentBar = ({
  data,
  height = '20px',
  tooltipOrientation = 'top',
  tooltipYOffset = -15,
}) => {
  const [chartWidth, setChartWidth] = useState(167);
  const [chartHeight, setChartHeight] = useState(24);
  const container = useRef(null);

  useEffect(() => {
    const resizeObserver = new ResizeObserver((event) => {
      setChartWidth(event[0].contentBoxSize[0].inlineSize);
      setChartHeight(event[0].contentBoxSize[0].blockSize);
    });

    if (container !== null) {
      resizeObserver.observe(container.current);
    }

    return () => {
      resizeObserver.disconnect();
    };
  });

  const dataValuesSum = data.reduce(
    (previous, current) => previous + current.value,
    0
  );

  return (
    <div ref={container} style={{ height, padding: 0 }}>
      <Chart
        showAxis={false}
        width={chartWidth}
        height={chartHeight}
        padding={0}
        containerComponent={<ChartContainer portalZIndex={110} />}
      >
        <ChartStack horizontal colorScale={data.map((point) => point.color)}>
          {data.map((point) => (
            <ChartBar
              key={point.title}
              data={[
                {
                  name: point.title,
                  x: '',
                  y: point.value,
                  label: point.label,
                },
              ]}
              barWidth={parseFloat(height) - 4}
              labelComponent={
                <ChartTooltip
                  orientation={tooltipOrientation}
                  dy={tooltipYOffset}
                  dx={tooltipXOffset(dataValuesSum, chartWidth, point.value)}
                />
              }
            />
          ))}
        </ChartStack>
      </Chart>
    </div>
  );
};

MultiSegmentBar.propTypes = {
  data: propTypes.array,
  height: propTypes.oneOfType([propTypes.number, propTypes.string]),
  tooltipOrientation: propTypes.oneOf(['top', 'bottom', 'right', 'left']),
  tooltipYOffset: propTypes.number,
};

export default MultiSegmentBar;
