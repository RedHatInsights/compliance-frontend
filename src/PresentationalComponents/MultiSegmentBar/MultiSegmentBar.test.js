import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MultiSegmentBar from './MultiSegmentBar';

const data = [
  {
    name: 'test1',
    color: '#fff',
    value: 5,
    label: 'test1',
  },
  {
    name: 'test2',
    color: '#000',
    value: 7,
    label: 'test2',
  },
];

describe('MultiSegmentBar', () => {
  it('expect to render without error', () => {
    const wrapper = shallow(<MultiSegmentBar data={data} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
