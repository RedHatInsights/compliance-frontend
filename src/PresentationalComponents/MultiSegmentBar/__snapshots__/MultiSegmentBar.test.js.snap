// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSegmentBar expect to render without error 1`] = `
<div
  style={
    Object {
      "height": "20px",
      "padding": 0,
    }
  }
>
  <Chart
    containerComponent={
      <ChartContainer
        portalZIndex={110}
      />
    }
    height={24}
    padding={0}
    showAxis={false}
    width={167}
  >
    <ChartStack
      colorScale={
        Array [
          "#fff",
          "#000",
        ]
      }
      horizontal={true}
    >
      <ChartBar
        barWidth={16}
        data={
          Array [
            Object {
              "label": "test1",
              "name": undefined,
              "x": "",
              "y": 5,
            },
          ]
        }
        labelComponent={
          <ChartTooltip
            dx={-34.791666666666664}
            dy={-15}
            orientation="top"
          />
        }
      />
      <ChartBar
        barWidth={16}
        data={
          Array [
            Object {
              "label": "test2",
              "name": undefined,
              "x": "",
              "y": 7,
            },
          ]
        }
        labelComponent={
          <ChartTooltip
            dx={-48.708333333333336}
            dy={-15}
            orientation="top"
          />
        }
      />
    </ChartStack>
  </Chart>
</div>
`;
